### ISSUE TITLE
network interface names are not consistently updated by udev rules

### ISSUE THREAD

**Issue by anderbubble** [warewulf/warewulf#926]
Captured from a Slack conversation:

> I have a cluster with a few nodes with the cables connected to the second eth port instead of the first one, so, meanwhile we cannot recable them I tried to set all nodes to eth0. The question is, that sometimes works but sometimes doesn't and renames it as eth1, even the udev is set to eth0. Do you know any idea what can be happen?

> A working node:

```
grep eth /etc/udev/rules.d/70-ww4-netname.rules
SUBSYSTEM=="net", ACTION=="add", ATTR{address}=="D8:5E:D3:69:3E:61", NAME="eth0"

ip l sh eth0
3: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether d8:5e:d3:69:3e:61 brd ff:ff:ff:ff:ff:ff

ip l sh eno1
2: eno1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether d8:5e:d3:69:3e:60 brd ff:ff:ff:ff:ff:ff
    altname enp98s0f0
```

> A non working node:

```
grep eth /etc/udev/rules.d/70-ww4-netname.rules                                                                                 
SUBSYSTEM=="net", ACTION=="add", ATTR{address}=="D8:5E:D3:69:34:0D", NAME="eth0" 

ip l sh eth0                                                                                                               
Device "eth0" does not exist.                                                                                                            

ip l sh eth1                                                                                                               
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000                                       
    link/ether d8:5e:d3:69:34:0d brd ff:ff:ff:ff:ff:ff                                                                                   

ip l sh eno1                                                                                                               
2: eno1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000                                       
    link/ether d8:5e:d3:69:34:0c brd ff:ff:ff:ff:ff:ff                                                                                   
    altname enp98s0f0 
```

> after a reboot now it is configured as eth0, but it is random when is set to eth0 or eth1

```
ip l sh eth0                                                                                                               
3: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000                               
    link/ether d8:5e:d3:69:34:0d brd ff:ff:ff:ff:ff:ff  
```

---

```
grep -A5 Source /etc/systemd/network/10-persistent-net.link
# Source: /var/lib/warewulf/overlays/wwinit/etc/systemd/network/10-persistent-net.link.ww

[Match]
MACAddress=D8:5E:D3:69:34:0D
[Link]
Name=eth0
```

---

> debuging some more this issue, I see that the igb drivers renames the interfaces, but sometimes udev tries to rename before and there is a conflict in the naming, I see that there are other net rules in /usr/lib/udev/rules.d/ with higher number in the names than warewulf udev rule (70-....). I'm not sure from where igb driver renames the interfaces, if gets the udev rules or another kind of data.

> is there a reason to set the net rule as 70- instead on higher or lower numbering?

**Comment by anderbubble** [comment]
The issue appears to occur when Warewulf is attempting to assign an interface name that is currently in use. Take the following situation:

- interface 0: eth0
- interface 1: eth1

Let's say we want to reverse these, and make interface 0 eth1, and interface 1 eth0. No matter what we do, if we rename interface 1 to eth0, it fails, because that name is already in use by interface 0, even though we intend to change it later. In this case, the same happens in reverse as well.

To avoid this in all cases, I think we'd need to first rename interfaces to a temporary name and then rename them to their intended final name. But I don't know if udev rules work that way.

**Comment by griznog** [comment]
At one point after I started using the network name (like I call my provisioning network `clusternet`) as teh network device name, I briefly looked at the udev rules template out of curiosity about naming my devices using the name of the network they were on, e.g. instead of `eth0` if a nic was on my provisioning network it'd be called `clusternet`. I mention this as a lead in to the question, do we have to use `ethX` as a pattern here? Since it seems like it'll always have the potential to conflict could we name it `ww0` instead?

FWIW, I'd still like to name mine after the network they are on, just haven't had tome to try to implement that.

**Comment by anderbubble** [comment]
> do we have to use ethX as a pattern here?

This isn't really a Warewulf decision either way; what to name the interface is an administrator decision. If other names were used, this issue should go away; but I expect many people will want to be able to use more typical names.

**Comment by griznog** [comment]
I was thinking of it as a safe as-installed default, rather than defaulting to something we know has the potential for conflicts.

**Comment by conxuro** [comment]
I saw another issue maybe related to this one, but with two interfaces renamed (eth0 and eth1).

The configurations for both interfaces are generated ok:
```
cat /etc/udev/rules.d/70-ww4-netname.rules.ww
[...]
SUBSYSTEM=="net", ACTION=="add", ATTR{address}=="7c:c2:55:26:dd:6c", NAME="eth0"                                                         
                                                                                                                                         
SUBSYSTEM=="net", ACTION=="add", ATTR{address}=="7c:c2:55:44:26:ee", NAME="eth1"                                                         

grep -E "IPADDR|HWADDR" /etc/sysconfig/network-scripts/ifcfg-default.conf                                                                             
IPADDR=192.168.131.39
HWADDR=7c:c2:55:26:dd:6c

grep -E "IPADDR|HWADDR"  /etc/sysconfig/network-scripts/ifcfg-storage.conf
IPADDR=192.168.130.39
HWADDR=7c:c2:55:44:26:ee
```

However, after boot, randomly sometimes it gets reverse renamed/configured (i.e. eth0 has the right IP but configured to the interface with the MAC address that should be eth1):
```
3: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 7c:c2:55:44:26:ee brd ff:ff:ff:ff:ff:ff
    inet 192.168.131.39/24 brd 192.168.131.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::7ec2:55ff:fe44:26ee/64 scope link
       valid_lft forever preferred_lft forever
4: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 7c:c2:55:26:dd:6c brd ff:ff:ff:ff:ff:ff
```

In this case I have not seen in the logs any message about "interface in use" when trying to rename, but I have no issues booting all the nodes if I set them with ethww or any name different from eth0,1.

**Comment by anderbubble** [comment]
@conxuro thanks for the case report.

Can you confirm what version of Warewulf you're running? I wonder if your issue, at least the inconsistency of it, may already be fixed in https://github.com/warewulf/warewulf/pull/1017.

**Comment by conxuro** [comment]
The version is 4.4.1-1, from github RPM:
```
wwctl version:   4.4.1-1.git_d6f6fed
rpc version: apiPrefix:"rc1" apiVersion:"1" warewulfVersion:"4.4.1-1.git_d6f6fed"
```

**Comment by anderbubble** [comment]
I believe this is as resolved as is feasible by #1017. There is still an issue when attempting to rename an interface to a name that is already in use; but this is an underlying limitation, not intrinsic to Warewulf.
